File: CodeLib/Graph/ZOJ 2314 Reactor Cooling.java
/*
 * ZOJ 2314 Reactor Cooling
 * Created by Rayn @@ 2014/08/10
 */
import java.io.*;
import java.util.*;

class Edge {
	int v, c, id;
	Edge r;
	
	public Edge(int v, int c) {
		this.v = v;
		this.c = c;
		this.id = -1;
		this.r = null;
	}
}
class NoSuchFlowException extends Throwable {}

class LimitFlow {
	int n;
	int source, sink;
	int[] f;
	ArrayList<ArrayList<Edge>> edge;
	
	public LimitFlow(int an, int[] s, int[] t, int[] b, int[] c) {
		n = an + 2;
		source = an;
		sink = an + 1;
		f = b.clone();
		edge = new ArrayList<ArrayList<Edge>>(n);
		for (int i = 0; i < n; ++i) {
			edge.add(new ArrayList<Edge>());
		}
		int[] d = new int[an];
		for (int i = 0; i < f.length; ++i) {
			Edge x = new Edge(t[i], c[i]-b[i]);
			Edge y = new Edge(s[i], 0);
			x.r = y;
			y.r = x;
			edge.get(s[i]).add(x);
			edge.get(t[i]).add(y);
			d[s[i]] -= b[i];
			d[t[i]] += b[i];
			x.id = i;
		}
		for(int i = 0; i < d.length; ++i) {
			if (d[i] > 0) {
				Edge x = new Edge(i, d[i]);
				Edge y = new Edge(source, 0);
				x.r = y;
				y.r = x;
				edge.get(source).add(x);
				edge.get(i).add(y);
			} else if (d[i] < 0) {
				Edge x = new Edge(sink, -d[i]);
				Edge y = new Edge(i, 0);
				x.r = y;
				y.r = x;
				edge.get(i).add(x);
				edge.get(sink).add(y);
			}
		}
	}
	public int[] work() throws NoSuchFlowException {
		int[] q = new int[n];
		int[] d = new int[n];
		int[] done = new int[n];
		int[] path = new int[n];
		while(true) {
			Arrays.fill(d, -1);
			q[0] = source;
			d[source] = 0;
			int begin = 0, end = 1;
			while (begin < end) {
				int u = q[begin++];
				for (Edge e : edge.get(u)) {
					if (e.c > 0 && d[e.v] == -1) {
						d[e.v] = d[u] + 1;
						q[end++] = e.v;
					}
				}
			}
			if(d[sink] == -1) {
				break;
			}
			Arrays.fill(done, 0);
			int len = 0;
			path[0] = source;
			while(len >= 0) {
				int back = path[len];
				if (back == sink) {
					int flow = Integer.MAX_VALUE;
					int todo = -1;
					for (int i = 0; i < len; ++i) {
						Edge e = edge.get(path[i]).get(done[path[i]]);
						if (e.c < flow) {
							flow = e.c;
							todo = i;
						}
					}
					for (int i = 0; i < len; ++i) {
						Edge e = edge.get(path[i]).get(done[path[i]]);
						e.c -= flow;
						e.r.c += flow;
					}
					len = todo;
				} else {
					while (done[back] < edge.get(back).size()) {
						Edge e = edge.get(back).get(done[back]);
						if (d[e.v] == d[back] + 1 && e.c > 0) {
							break;
						} else {
							++done[back];
						}
					}
					if (done[back] == edge.get(back).size()) {
						d[back] = -1;
						--len;
					} else {
						path[++len] = edge.get(back).get(done[back]).v;
					}
				}
			}
		}
		for (Edge e : edge.get(source)) {
			if (e.c > 0) {
				throw new NoSuchFlowException();
			}
		}
		for (ArrayList<Edge> ae : edge) {
			for (Edge e : ae) {
				if(e.id != -1) {
					f[e.id] += e.r.c;
				}
			}
		}
		return f;
	}
}
public class Main {
	
	private static Scanner cin;
	//public static FileInputStream fin;

	public static void main(String[] args) throws FileNotFoundException {
		//fin = new FileInputStream("D:/Code/in.txt");
		//System.setIn(fin);
		cin = new Scanner(System.in);
		
		int T = cin.nextInt();
		for (int ri = 1; ri <= T; ++ri) {
			if(ri > 1) {
				System.out.println();
			}
			int n = cin.nextInt();
			int m = cin.nextInt();
			int[] s = new int[m];
			int[] t = new int[m];
			int[] b = new int[m];
			int[] c = new int[m];
			for (int i = 0; i < m; ++i) {
				s[i] = cin.nextInt() - 1;
				t[i] = cin.nextInt() - 1;
				b[i] = cin.nextInt();
				c[i] = cin.nextInt();
			}
			LimitFlow Solver = new LimitFlow(n, s, t, b, c);
			try {
				int[] res = Solver.work();
				System.out.println("YES");
				for (int f : res) {
					System.out.println(f);
				}
			} catch (NoSuchFlowException e) {
				System.out.println("NO");
			}
		}
	}
}

File: CodeLib/Graph/ZOJ 2314 Reactor Cooling.cpp
/*
** ZOJ 2314 Reactor Cooling
** Created by Rayn @@ 2014/08/10
*/
#include <cstdio>
#include <cstring>
#include <cmath>
#include <map>
#include <vector>
#include <set>
#include <queue>
#include <algorithm>
using namespace std;
const int maxn = 250;
const int maxm = 250*250*2;
const int INF = 0x3F3F3F3F;

struct Edge
{
    int from, to, cap, flow;
    int next;

    Edge() {}
    Edge(int from, int to, int cap, int flow, int next):
        from(from), to(to), cap(cap), flow(flow), next(next) {}
};

struct LimitFlow
{
    int n, m, sz;
    int source, sink;
    int b[maxm], c[maxm], limit[maxn];
    int d[maxn], head[maxn], cur[maxn];
    bool vis[maxn];
    Edge edge[maxm];

    void Init(int n, int m)
    {
        this->n = n;
        this->m = m;
        sz = 0;
        source = 0;
        sink = n + 1;
        memset(head, -1, sizeof(head));
        memset(limit, 0, sizeof(limit));
    }
    void AddEdge(int from, int to, int cap, int flow)
    {
        edge[sz] = Edge(from, to, cap, flow, head[from]);
        head[from] = sz++;
        cur[from] = head[from];
    }
    void ReadGraph()
    {
        for(int i = 0; i < m; ++i)
        {
            int u, v;
            scanf("%d%d%d%d", &u, &v, &b[i], &c[i]);
            AddEdge(u, v, c[i]-b[i], 0);
            AddEdge(v, u, 0, 0);
            limit[u] -= b[i];
            limit[v] += b[i];
        }
    }
    void Work()
    {
        int in = 0;
        for(int i = 1; i <= n; ++i)
        {
            if (limit[i] > 0) {
                AddEdge(source, i, limit[i], 0);
                AddEdge(i, source, 0, 0);
                in += limit[i];
            } else if (limit[i] < 0) {
                AddEdge(i, sink, -limit[i], 0);
                AddEdge(sink, i, 0, 0);
            }
        }
        if(MaxFlow(source, sink) != in)
        {
            printf("NO\n");
            return ;
        }
        printf("YES\n");
        for(int i = 0; i < m; ++i)
        {
            printf("%d\n", edge[2*i].flow + b[i]);
        }
    }
    bool bfs()
    {
        memset(vis, false, sizeof(vis));
        queue<int> Q;
        Q.push(source);
        vis[source] = true;
        d[source] = 0;
        while(!Q.empty())
        {
            int u = Q.front();
            Q.pop();
            for(int i = head[u]; i != -1; i = edge[i].next)
            {
                int v = edge[i].to;
                int w = edge[i].cap - edge[i].flow;
                if(!vis[v] && w > 0)
                {
                    vis[v] = true;
                    Q.push(v);
                    d[v] = d[u] + 1;
                }
            }
        }
        return vis[sink];
    }
    int dfs(int u, int a)
    {
        if(u == sink || a == 0)
            return a;
        int flow = 0, f;
        for(int& i = cur[u]; i != -1; i = edge[i].next)
        {
            int v = edge[i].to;
            int w = edge[i].cap - edge[i].flow;
            if(w > 0 && d[v] == d[u] + 1 && (f = dfs(v, min(a, w))) > 0)
            {
                edge[i].flow += f;
                edge[i^1].flow -= f;
                flow += f;
                a -= f;
                if(a == 0)
                    break;
            }
        }
        return flow;
    }
    int MaxFlow(int st, int ed)
    {
        int flow = 0;
        while(bfs())
        {
            memcpy(cur, head, sizeof(head));
            flow += dfs(st, INF);
        }
        return flow;
    }
};


LimitFlow Solver;

int main()
{
#ifdef _Rayn
    freopen("in.txt", "r", stdin);
#endif
    int T, N, M;
    scanf("%d", &T);
    while(T--)
    {
        scanf("%d%d", &N, &M);
        Solver.Init(N, M);
        Solver.ReadGraph();
        Solver.Work();
        if(T > 0)
            puts("");
    }
    return 0;
}
